Mike o user search by date and status

MikeOlivas opened the pull request:
- Updated userManagement view to include two text inputs, two drop-down selectors, a dateRangePicker, and a "search" and "clear" button. All inputs play nicely with each other and retain the previous text to support sequential searches - so an admin can drill down into the results and whittle away users they're not interested in. Can now search by:
  1. Badge number
  2. Name (caps insensitive; allows partial match)
  3. Status {User, Manager, Admin}
  4. Registration date range
  5. Permissions (users with access to a particular station) [I just piggy-backed on the work Bishoy did]


- Updated post route for userManagement in index to find (and return) the intersection of the various inputs requested (if and when they exist).

- Some other little things.

MikeOlivas commented on routes/index.js:
```diff
@@ -579,3 +658,14 @@ function checkRegistration(req, res, next) {
     next();
   }
 }
+
```
I was thinking checkRegistration might be able to call this as a helper function above; I couldn't use that function directly because in my case, I always want to preserve the data captured. checkRegistration only preserves data if/when there are errors. I'd like to worry about that later though, if possible so we can get this functionality included sooner rather than later.

danieleynis commented on routes/index.js:
```diff
@@ -298,26 +379,25 @@ module.exports = function (passport) {
+        dbAPI.modifyUser(badgeID, { status: "Manager" }).then(function (result) {
           if (result == undefined) {
-            console.log("Error");
+            flash("error", "Internal error: unable to promote " + badgeID);
```
Shouldn't this be `req.flash()`?

MikeOlivas replied:
Whoops! Definitely. I'll fix that real quick.

MikeOlivas replied:
Done.

MikeOlivas commented:
I see why the 8d printer thing was happening - that's a quick fix. I had originally figured they'd click the "clear" button if they wanted remove the stations chosen... since that is the default behavior of the selector when no value is picked. However, I can see the value in supporting an "All stations" choice. I'll work on that.

Unfortunately, I won't have time to look into asnchronous JQuery this weekend though. This is literally all I've done all weekend and I'm starting to freak out about my classes a little bit :-/ I have 8 hours to try and get some of my homework done. I'll see if I can check in those other two fixes before I get started though.

danieleynis commented:
That's fine, I am not saying you have to get it done this weekend. Not sure how much time you have, but we can work through it this week and next weekend. I was gonna use the async jquery requests for updating user information, so perhaps that would be an example you could take a look at once I am done with it. For now yeah just focus on those fixes I mentioned and don't worry about the jquery stuff for now.

MikeOlivas commented:
Oh that's good; I just don't want you guys to think I'm sitting on my hands or anything. I feel guilty for every second I spend away from this project. LOL I'll look for your example in the coming week. I do appreciate the feedback though - that's why we do these pull requests, right? Four sets of eyes are better than one.

I have updated the two requested changes. I went ahead and added an "All" option for the status selector too. Since you mentioned it, I think its a good idea... now each individual search control can be cleared individually or the "clear" button can be clicked to clear all cached search inputs.

MikeOlivas commented:
I just thought of another option rather than re-direct too... we could always factor out the middleware for get and post into named functions (like I did before with my alerts setup). Then, after the post does the heavy lifting, it can just call next(). We pass a flag in the req.body (to signal that it already has the user and station array(s)). Just an idea I wanted to put on paper while I was thinking it.

danieleynis commented:
@MikeOlivas I checked you changes and it looks good. I think we just need to resolve that "Confirm Form Resubmission" error when clicking the back button on the browser. Let's talk about it at today's meeting and figure something out.

danieleynis commented:
Did some searching and ran across this: https://stackoverflow.com/questions/18614986/confirm-form-resubmission-issue-on-page-rendered-thru-app-post-in-express
Something we could look into

danieleynis commented on routes/index.js:
```diff
@@ -196,35 +196,140 @@ module.exports = function (passport) {
+            data.users = stationMatchesArray;
+          }
+          
           res.render('userManagement.njk', data);
```
I think the problem for the Confirm Form Re submission is still there is because you are trying to render the userManagement page in the post. Essentailly what should happen is that this post returns a JSON object with all the matched users according to given search criteria.

danieleynis commented on views/userManagement.njk:
```diff
@@ -16,88 +16,179 @@
+
+
+  // Attach a submit handler to the searchSubmit button
+  $('#searchSubmit').submit("click", function(event) {
```
Im confused why do you need to pass "click" here?

danieleynis commented on views/userManagement.njk:
```diff
@@ -16,88 +16,179 @@
+      status = $form.find("input[statusSelector]").val(),
+      badge = $form.find("input[badgeInput]").val(),
+      station = $form.find("input[stationSelector]").val(),
+      url = $form.attr("POST");
```
wouldnt you need to retrieve the "action" attribute from the form for the url?

MikeOlivas commented:
I think that's the behavior Daniel was expecting, since we're not rendering the page any more. The result of clicking "back" from the edit details button is the equivalent of doing another "get" - which populates all of the users. I think the search fields should probably be cleared out also though to make that apparent. @danieleynis , can you confirm that is true?

If so, I'll work on making sure the input fields are cleared when the "edit details" button is clicked so that when the user clicks back from the edit details page he/she has blank search fields and a list of all the users.

MikeOlivas commented:
@BishoyHanna91 @danieleynis I pushed a change for the above to show what I mean. It's an easy change to undo if I'm totally off-base.

danieleynis commented:
@MikeOlivas @BishoyHanna91 hmmm that's a good thing to consider. I think perhaps we want to leave the previous search configuration after they hit back since I feel like it would be annoying to have it cleared every time. Would it be possible after they hit back to have it immediately do the search with the previous search configuration. Hope that make sense. Lmk.

MikeOlivas commented:
Not that I can think off off-hand. Since the JQuery intercepts the post, it’s like it never happened. When you click back in the debugger, it immediately goes to the “get” route.

danieleynis commented:
In you latest change you put that in to clear the search when the user hits back right? If that is the case when the document is ready could you use jQuery to trigger a click on the button?

MikeOlivas commented:
That’s a good idea. I’ll log out of Linux and into my Windows machine to give it a go. I’ll Report back shortly.

MikeOlivas commented:
@danieleynis You sir, are a genius. Pushing the update momentarily.

danieleynis commented:
I think this looks good to me. Perhaps @BishoyHanna91 you can take a look over the code one last time, just to make sure I did not miss anything.

BishoyHanna91 commented:
Unfortunately I still see the same problem. After pressing the back button, every attribute  is kept in the filter bars as status: Manager , station: whatever, but it returns all the users again which the user will need to press the search button again to do the filtration process again.
@danieleynis  I am not sure is this a major problem we need to worry about or what is done is OK

danieleynis commented:
This is the intended behaviour now. Not sure if you following my chat with Mike. I think it's good for now. I would say go ahead and merge @MikeOlivas.

MikeOlivas commented:
Okay, I will go ahead and merge then @danieleynis. @BishoyHanna91 are you on campus today? I'm here briefly in the Capstone room.
